### ISSUE TITLE
As an user, I would like to take a picture of my handwritten notes or printed documents, scan them and upload it directly to my ownCloud

### ISSUE THREAD

**Issue by davigonz** [owncloud/android#2119]
The current ownCloud Android app has different options to upload files to an ownCloud server, such as selecting them from the internal storage, selecting content from other apps or uploading photos directly from the camera.

It would be great to include a new option which enables the user to take a new picture with the camera and select the edges of handwritten notes or printed documents. After that, the image will be processed, adjusted to a better view and uploaded to a folder on the server.

**Comment by sd1998** [comment]
@davigonz I would like to work on this.

**Comment by davigonz** [comment]
Great @sd1998 , glad to hear you want to contribute with this. Just for you to know, there's a bounty for implementing this feature.

**Comment by sd1998** [comment]
@davigonz Where can I get the information about the bounty for this feature?

**Comment by michaelstingl** [comment]
@sd1998 https://www.bountysource.com/issues/55217024

**Comment by sd1998** [comment]
@michaelstingl  Thanks.

**Comment by sd1998** [comment]
@davigonz  For this feature should I use a 3rd party dependency or should I try to implement the feature from scratch using opencv etc.?

**Comment by Unsigno** [comment]
Hi guys , when you talk about "the image will be processed", do you mean using OCR to get plain text or adjust the marked edges to get a rect and aligned document ?

**Comment by davigonz** [comment]
> Hi guys , when you talk about "the image will be processed", do you mean using OCR to get plain text or adjust the marked edges to get a rect and aligned document ?

Actually Is not OCR itself, is just about taking a picture and adding some kind of filters so that the picture seems a scanned document.

**Comment by Unsigno** [comment]
Oki , i am looking for a PR and have some questions.

1 - Could be added OpenCV as dependency to do this ?
2 - If i send a PR, it has to contain all the translations of the texts used ?

**Comment by sd1998** [comment]
@Unsigno I am already working on this. Could you work on any other issue? Thanks.

**Comment by Unsigno** [comment]
@sd1998  You could mark it on bountysource as started and give an aproximated time to prevent bounty hunters :p . give some news if you leave it please .

If some one can asnwer 2 i still curious .

**Comment by davigonz** [comment]
> 1 - Could be added OpenCV as dependency to do this ?
> 2 - If i send a PR, it has to contain all the translations of the texts used ?

@Unsigno 

1. Yes, OpenCV could be used to detect the document edges for example.
2. Not necessarily, including texts in english is enough since they will be automatically translated by transifex. If you have a look at the master commits history, you will see that many of them are automatically added by owncloud-bot and include the translations I mention.

**Comment by sd1998** [comment]
@davigonz I am using camera2 API it requires the API level 21 and currently, the minimum API level has been set to 14. Should I change it? Or try an alternate approach?

**Comment by davigonz** [comment]
> I am using camera2 API it requires the API level 21 and currently, the minimum API level has been set to 14. Should I change it? Or try an alternate approach?

@sd1998 

Were not you using [https://github.com/jhansireddy/AndroidScannerDemo](url) to implement this feature? What is camera2 API used for?

If there's no way to include this feature in APIs lower than 21, I would include a condition to disable the feature for versions lower than 21 instead of setting the minimum API level. 

Keep in mind we still have users with Android versions previous to Lollipop (API 21).

**Comment by sd1998** [comment]
@davigonz I did not use https://github.com/jhansireddy/AndroidScannerDemo but have implemented the feature in a better way. And it will also work for minSDk = 14 thus supporting Android version previous to Lollipop.

**Comment by Unsigno** [comment]
Hi all, this issue is still open ?

**Comment by davigonz** [comment]
> Hi all, this issue is still open ?

Yes @Unsigno , but it was delayed because is not trivial at all, have a look at [this overview](https://github.com/owncloud/android/pull/2143#issuecomment-395018491).

**Comment by iamzubin** [comment]
hey, so will this issue be on google summer of code this year?
as this issue has a bounty as well.
I can do it as a GSoC project or I can fix it a now,

**Comment by michaelstingl** [comment]
@iamzubin previous attempt: https://github.com/owncloud/android/pull/2143

**Comment by michaelstingl** [comment]
No plans for GSoC yet, but yes. there's an open bounty:
https://www.bountysource.com/issues/55217024

**Comment by Bables55** [comment]
The iOS version is able to achieve this feature. Is it difficult to adopt this feature in the Android version?

**Comment by hannesa2** [comment]
> The iOS version is able to achieve this feature. Is it difficult to adopt this feature in the Android version?

To make it perfect: yes

I made such a feature, which fulfill my needs https://github.com/hannesa2/owncloud-android/releases. 
But it's not perfect. 

@michaelstingl 
About the bounty. Ok, it's not that much, but why not ask for it ?
What has to be fulfilled to receive it ? I ask because I know, there will be always improvements and there is alwasy someone who is not satisfied ...

**Comment by vibhuti019** [comment]
Hello, can I work on this issue? This is a really old issue, is this still open? I would like to work on it.

**Comment by hannesa2** [comment]
> Hello, can I work on this issue? This is a really old issue, is this still open? I would like to work on it.

@vibhuti019 Please see my post, my fork already comes with that feature

**Comment by vibhuti019** [comment]
@hannesa yes, I saw that now,  I was working on this issue..   @abelgardep please let me know if you wish to implement some changes in it. I would love to upgrade it ..

**Comment by fesave** [comment]
Hi @hannesa2 ! I have tried your fork of ownCloud with the integrated document scanner, it works really well, but I have a question. I have not been able to get the Scan edge document to work, it always throws the same error: Use darker background. Can you tell me how it works?

Also, we have seen that the size of the .apk increases excessively and it's a bit annoying the automatic detection and upload of the scanned document, since sometimes it doesn't detect correctly and uploads a single color image. Do you have any improvement or update in mind? Thanks and good work!

**Comment by fesave** [comment]
After an intense search, I have found a couple of possible solutions to evaluate.

---

The first one is not free, called Scanbot SDK and is available for Android and iOS.

- Advantages: 
  - It is very well documented.
  - It is continuously updated.
  - Fixed price.
  - Support.
  - [Demo App](https://play.google.com/store/apps/details?id=io.scanbot.sdk.showcase) where we can test it. (It works fine!)
  
- Cons:
   - We have to pay .
   - We need to talk with a sales manager.
   - I don't know if its code can be part of an open source project like ownCloud.

More info [here](https://scanbot.io/)

---

The second one is free, it is a Github's library.

- Advantages: 
  - It is free (obviously).
  - It uses CameraX and OpenCV.
  - Made using Kotlin.

- Cons:
   - Some bugs.
   - Will it be deprecated in the future?.
 
More info [here](https://github.com/Kuama-IT/android-document-scanner)

---
The last one is free too!, it is a Github's library.

- Advantages: 
  - It is free (obviously).
  - It uses CameraX and tiny version of OpenCV.
  - Made using Kotlin.

- Cons:
   - Some bugs.
   - Will it be deprecated in the future?.

More info [here](https://github.com/zynkware/Document-Scanning-Android-SDK)

**Comment by JuancaG05** [comment]
New Google API to scan documents, available for API 21 or above:
- https://android-developers.googleblog.com/2024/02/ml-kit-document-scanner-api.html
- https://developers.google.com/ml-kit/vision/doc-scanner

**Comment by BWWIT** [comment]
Uhh...no progress since 2021 :-(

Keep the requirements at a minimum to start with or you "drown".

**Need-to have:** 
- take a picture and upload directly to webdav folder (pdf would be nice, but not a must-have).

**Nice-to-have:**
- document border adjustment
- image filter / optimization
- other fancy stuff

**Comment by jesmrec** [comment]
yep, that's an ancient feature request, not implemented yet for several reasons that are not in our hands to change 😄 

If someone is brave enough to take a look into [this](https://github.com/owncloud/android/issues/2119#issuecomment-1978306854)  and make a proposal, we'll be happy to check it.
